### PR TITLE
DTSRD-2138.Delete DX address for an organisation

### PR DESCRIPTION
**Pull request by SabinaHMCTS**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/DTSRD-2138

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java [hunk 1]
```diff
@@ -208,6 +209,18 @@ public void editPbaScenarios() {
...
+    public void deleteDxAddressShouldReturnSuccess() {
+        log.info("deleteDxAddressShouldReturnSuccess :: STARTED");
+        String orgIdentifier = createAndUpdateOrganisationToActive(hmctsAdmin);
+        JsonPath jsonPath = professionalApiClient.retrieveOrganisationDetails(orgIdentifier, hmctsAdmin, OK);
+        String dxNumber = jsonPath.get("contactInformation[0].dxAddress[0].dxNumber");
+        String dxExchange = jsonPath.get("contactInformation[0].dxAddress[0].dxExchange");
+        professionalApiClient.deleteDXAddress(new DxAddressCreationRequest(dxNumber, dxExchange),
+                orgIdentifier, hmctsAdmin, NO_CONTENT);
+        log.info("deleteDxAddressShouldReturnSuccess :: END");
     }
```
Comment (SabinaHMCTS) [inline src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:225]: this is not a test .. need @Test above the method also run the test and make it pass add assert statements .check that the test covers all functionality by run with code coverage

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java [hunk 2]
```diff
@@ -208,6 +209,18 @@ public void editPbaScenarios() {
     public void deleteOrganisationScenarios() {
         deletePendingOrganisationShouldReturnSuccess();
         deleteActiveOrganisationShouldReturnSuccess();
+        deleteDxAddressShouldReturnSuccess();
+    }
+
```
Comment (SabinaHMCTS) [inline src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:214]: add negative tests to cover exceptions from eg empty contact lists empty dx addresses etc

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java [hunk 3]
```diff
@@ -208,6 +213,75 @@ public void editPbaScenarios() {
...
+        String orgIdentifier = createAndUpdateOrganisationToActive(hmctsAdmin, orgCreationRequest);
+
+        Map<String, Object> deleteDxAddressResponse = professionalApiClient.deleteDxAddressWithResponse(
+                new DxAddressCreationRequest("DX 1234567890", "dxExchange"),orgIdentifier);
+        assertThat((String) deleteDxAddressResponse.get("errorDescription")).contains("No contact information  found");
+        log.info("deleteEmptyContactListShouldReturnError :: END");
+    }
+
+    @Test
+    public void deleteEmptyDxAddressShouldReturnError() {
```
Comment (rajkatla-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:257]: as discussed, rename this method to deleteNonExistDxAddressShouldReturnError

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/DeleteDxAddressTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        Map<String, Object> deleteResponse = deleteDxAddress(new DxAddressCreationRequest("invalidDxNumber",
+                "dxExchange"));
+        assertThat(deleteResponse.get("http_status")).isEqualTo(400);
+    }
+
+    private Map<String, Object> deleteDxAddress(DxAddressCreationRequest dxAddressCreationRequest) {
+        OrganisationCreationRequest organisationCreationRequest = organisationRequestWithAllFields().build();
+        String orgIdentifier = createAndActivateOrganisationWithGivenRequest(organisationCreationRequest);
+        return professionalReferenceDataClient.deleteDxAddress(hmctsAdmin, orgIdentifier, dxAddressCreationRequest);
+    }
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/DeleteDxAddressTest.java:34]: Please these 2 tests in IT test
1) add empty contact information list error scenario
2) non exist dx address error scenario

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/DeleteDxAddressTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Test
+    void returns_400_when_deleting_non_exist_dx_number() {
+        OrganisationCreationRequest organisationCreationRequest = someMinimalOrganisationRequest().build();
+        String orgIdentifier = createAndActivateOrganisationWithGivenRequest(organisationCreationRequest);
+
+        Map<String, Object> deleteResponse = professionalReferenceDataClient.deleteDxAddress(hmctsAdmin, orgIdentifier,
+                 new DxAddressCreationRequest("DX 1234567890","dxExchange"));
+
+        assertThat(deleteResponse.get("http_status")).isEqualTo(400);
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/DeleteDxAddressTest.java:53]: api return status code 404 when dx address is not found

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java [hunk 6]
```diff
@@ -697,6 +703,61 @@ public ResponseEntity<OrganisationEntityResponse> retrieveOrganisationByUserId(
...
+        description = "Internal Server Error",
+        content = @Content
+    )
+
+    @DeleteMapping(path = "/dxAddress/{orgId}")
+    @ResponseStatus(value = HttpStatus.NO_CONTENT)
+    @Secured({"prd-admin"})
+    public void deleteDxAddressOfOrganisation(
+        @io.swagger.v3.oas.annotations.parameters.RequestBody(description = "deletedxAddressRequest")
+        @Valid @NotNull @RequestBody DxAddressCreationRequest dxAddressRequest,
```
Comment (rajkatla-hmcts) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:754]: The DxAddressCreationRequest class has both dxNumber and dxExchange properties as mandatory. However, for deleting an dx address, only dxNumber is used. Can you rename Request to a more generic name, like DxAddressRequest, as the current name is confusing for delete operations? Additionally, please remove the @NotNull annotation on the dxExchange property. For the old logic, manually handle the null check for the dxExchange field and return an error where necessary.

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]
